airflow/hooks/hive_hooks.py

```diff
@@ -621,17 +621,34 @@ def max_partition(self, schema, table_name, field=None, filter=None):
         >>> hh.max_partition(schema='airflow', table_name=t)
         '2015-01-01'
         """
-        parts = self.get_partitions(schema, table_name, filter)
-        if not parts:
-            return None
-        elif len(parts[0]) == 1:
-            field = list(parts[0].keys())[0]
-        elif not field:
+        self.metastore._oprot.trans.open()
+        table = self.metastore.get_table(dbname=schema, tbl_name=table_name)
+        if len(table.partitionKeys) != 1:
             raise AirflowException(
-                "Please specify the field you want the max "
-                "value for")
+                "The table isn't partitioned by a single partition key")
+
+        key_name = table.partitionKeys[0].name
+        if field is not None and key_name != field:
+            raise AirflowException("Provided field is not the partition key")
+
+        part_names = self.metastore.get_partition_names(schema,
+                                                        table_name,
+                                                        max_parts=32767)
```

Comment:
max_parts shouldn't be a magic number, and there should be an explanation why this number is being used (and potentially make it configurable).

Reply:
I'll add same comment as in get_partitions() to explain the number. Probably not going to make it configurable as it does not give too much value to get max_partition on a random set of partitions.

Reply:
Sounds good, let's make it a class constant then.

Reply:
Agreed. In this case you are limited by the Java API for the metastore which limits this to a short int.

From the Javadocs : https://hive.apache.org/javadocs/r1.2.2/api/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.html#listPartitions(java.lang.String,%20java.lang.String,%20java.util.List,%20short)